src/mewbot/api/type_guard.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-FileCopyrightText: 2023 Mewbot Developers <mewbot@quicksilver.london>
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+"""
+Type guard exists to provide methods for run time typing validation.
+
+E.g. a decorator which checks that an InputEvent being passed to a function is of appropriate
+     type.
+
+The intent of the tools here is to provide a more elegant alternative to constructs like
+
+.. code-block:: python
+
+    def matches(self, event: InputEvent) -> bool:
+        if not isinstance(event, DiscordMessageCreationEvent):
+            return False
+        [Function body]
+
+which could be found in a Trigger.
+
+Instead, using one of the decorators provided in this module, you might write
+
+
+.. code-block:: python
+
+    @check_input_event_against_sig_types
+    def matches(self, event: InputEvent) -> bool:
+        [Function body]
+
+Or, as a more full example
+
+.. code-block:: python
+
+    # Demo Code
+
+    from mewbot.io.discord import DiscordInputEvent, DiscordUserJoinInputEvent
+    from mewbot.io.http import IncomingWebhookEvent
+
+    class A(Trigger):
+
+        @staticmethod
+        def consumes_inputs() -> set[type[InputEvent]]:
+            return {DiscordInputEvent, IncomingWebhookEvent}
+
+        @check_input_event_against_sig_types
+        def matches(self, event: DiscordInputEvent | IncomingWebhookEvent) -> bool:
+            return True
+
+    print(A.matches) # <function A.matches at 0x7f6e703f4a40>
+    print(A.matches.__doc__) # <function A.matches at 0x7f6e703f4a40>
+    print(A().matches(InputEvent())) # False
+    print(A().matches(DiscordInputEvent())) # True
+    print(A().matches(DiscordUserJoinInputEvent("[some text]"))) # True
+
+"""
+
+import functools
+
+from typing import Callable, TypeVar, get_type_hints, Union
+
+from mewbot.api.v1 import Trigger, Condition, InputEvent
+
+Component = TypeVar("Component", bound=Union[Trigger, Condition])
+Event = TypeVar("Event", bound=InputEvent)
+
+
+def check_input_event_against_sig_types(
```

Comment:
I'm ok with this name, but if you're still unsure (from the commit messages) another naming approach might be along the lines of `pre_filter_non_matching_events`?

Reply:
`pre_filter_non_matching_events` sounds good to me! I'll make the change.